python/pyarrow/tests/test_parquet.py

```diff
@@ -632,6 +632,18 @@ def make_sample_file(table_or_df):
     return pq.ParquetFile(buf)
 
 
+def test_compression_level():
+    arr = pa.array(list(map(int, range(1000))))
+    data = [arr, arr]
+    table = pa.Table.from_arrays(data, names=['a', 'b'])
+    _check_roundtrip(table, expected=table, compression="gzip",
+                     compression_level=1)
+    _check_roundtrip(table, expected=table, compression="gzip",
+                     compression_level=5)
+    _check_roundtrip(table, expected=table, compression="gzip",
+                     compression_level=[{'a': 2, 'b': 3}])
```

Comment:
Can you add a test to exhibit whether an error from an invalid compression level bubbles up to Python?

Reply:
Yes, this was a good idea since I noticed that the zlib codec lazy initialization in Arrow can result into an assert if an invalid compression level is specified.
I added a patch for arrow-cpp to accomplish two things:
 * Return an error if the codec doesn't support a compression level instead of ignoring it silently. This could help avoid confusion if a user doesn't see any change in compression ratio.
 * Make sure that Codec::Create also initializes the codec by calling the newly added method `Status Init()`.
 * Add a test